subprojects/base-services/src/main/java/org/gradle/internal/service/InjectUtil.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.service;
+
+import javax.inject.Inject;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Modifier;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+class InjectUtil {
+
+    /**
+     * Selects the single injectable constructor for the given type.
+     * The type must either have only one public or package-private default constructor,
+     * or it should have a single non-private constructor annotated with {@literal @}{@link Inject}.
+     *
+     * @param type the type to find the injectable constructor of.
+     */
+    static Constructor<?> selectConstructor(Class<?> type) {
+        if (isInnerClass(type)) {
+            throw new ServiceValidationException(
+                String.format(
+                    "Unable to select constructor for non-static inner class %s.",
+                    format(type)
+                )
+            );
+        }
+
+        List<Constructor<?>> constructors = Arrays.asList(type.getDeclaredConstructors());
```

Comment:
This is performance critical code, so reducing the number of objects created is important here.

I would keep around the array and access it directly below.